### PR TITLE
build: merge multiple features specified at once

### LINKED ISSUE #24588: ARMv8 CPU features management is broken for some cases
### System Information

Platform: amrv8 + dotprodoct extension or fp16 neon extension
OS: Linux
Compiler: GCC.

### Detailed description

OpenCV handles ARMv8 extensions incorrectly. Other architectures use extra flags to enable extensions, e.g. `msse3`, `-mavx`. ARMv8 defines extra architecture options in `march` as  `-march=armv8.2-a+fp16+bf16`.

`cmake -DCPU_BASELINE="NEON;NEON_FP16;NEON_DOTPROD" ../opencv`
Substitutes wrong flags to compiler: `-march=armv8.2-a+dotprod -march=armv8.2-a+fp16`

It leads to build error:
```
[ 57%] Building CXX object modules/core/CMakeFiles/opencv_core.dir/src/matrix.cpp.o
In file included from /home/aromanov/opencv/modules/core/include/opencv2/core/cv_cpu_dispatch.h:83,
                 from /home/aromanov/opencv/modules/core/include/opencv2/core/cvdef.h:361,
                 from /home/aromanov/opencv/modules/core/include/opencv2/core.hpp:52,
                 from /home/aromanov/opencv/modules/core/include/opencv2/core/utility.hpp:56,
                 from /home/aromanov/opencv/modules/core/src/precomp.hpp:53,
                 from /home/aromanov/opencv/modules/core/src/matmul.dispatch.cpp:44:
/usr/lib/gcc/aarch64-linux-gnu/9/include/arm_neon.h: In function ‘cv::hal_baseline::v_uint32x4 cv::hal_baseline::v_dotprod_expand(const cv::hal_baseline::v_uint8x16&, const cv::hal_baseline::v_uint8x16&, const cv::hal_baseline::v_uint32x4&)’:
/usr/lib/gcc/aarch64-linux-gnu/9/include/arm_neon.h:33666:1: error: inlining failed in call to always_inline ‘uint32x4_t vdotq_u32(uint32x4_t, uint8x16_t, uint8x16_t)’: target specific option mismatch
33666 | vdotq_u32 (uint32x4_t __r, uint8x16_t __a, uint8x16_t __b)
      | ^~~~~~~
In file included from /home/aromanov/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:221,
                 from /home/aromanov/opencv/modules/core/src/precomp.hpp:88,
                 from /home/aromanov/opencv/modules/core/src/matmul.dispatch.cpp:44:
/home/aromanov/opencv/modules/core/include/opencv2/core/hal/intrin_neon.hpp:708:37: note: called from here
  708 | OPENCV_HAL_IMPL_NEON_DOT_PRODUCT_OP(v_uint32x4, v_uint8x16, u32)
      |                                     ^
/home/aromanov/opencv/modules/core/include/opencv2/core/hal/intrin_neon.hpp:705:12: note: in definition of macro ‘OPENCV_HAL_IMPL_NEON_DOT_PRODUCT_OP’
  705 |     return _Tpvec1(vdotq_##suffix(c.val, a.val, b.val)); \
      |            ^~~
In file included from /home/aromanov/opencv/modules/core/include/opencv2/core/cv_cpu_dispatch.h:83,
                 from /home/aromanov/opencv/modules/core/include/opencv2/core/cvdef.h:361,
                 from /home/aromanov/opencv/modules/core/include/opencv2/core.hpp:52,
                 from /home/aromanov/opencv/modules/core/include/opencv2/core/utility.hpp:56,
                 from /home/aromanov/opencv/modules/core/src/precomp.hpp:53,
                 from /home/aromanov/opencv/modules/core/src/matmul.dispatch.cpp:44:
/usr/lib/gcc/aarch64-linux-gnu/9/include/arm_neon.h:33666:1: error: inlining failed in call to always_inline ‘uint32x4_t vdotq_u32(uint32x4_t, uint8x16_t, uint8x16_t)’: target specific option mismatch
33666 | vdotq_u32 (uint32x4_t __r, uint8x16_t __a, uint8x16_t __b)
      | ^~~~~
In file included from /home/aromanov/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:221,
                 from /home/aromanov/opencv/modules/core/src/precomp.hpp:88,
                 from /home/aromanov/opencv/modules/core/src/matmul.dispatch.cpp:44:
/home/aromanov/opencv/modules/core/include/opencv2/core/hal/intrin_neon.hpp:708:37: note: called from here
  708 | OPENCV_HAL_IMPL_NEON_DOT_PRODUCT_OP(v_uint32x4, v_uint8x16, u32)
      |                                     ^
/home/aromanov/opencv/modules/core/include/opencv2/core/hal/intrin_neon.hpp:705:12: note: in definition of macro ‘OPENCV_HAL_IMPL_NEON_DOT_PRODUCT_OP’
  705 |     return _Tpvec1(vdotq_##suffix(c.val, a.val, b.val)); \
      |            ^~~~~
make[3]: *** [modules/core/CMakeFiles/opencv_core.dir/build.make:597: modules/core/CMakeFiles/opencv_core.dir/src/matmul.dispatch.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [CMakeFiles/Makefile2:2297: modules/core/CMakeFiles/opencv_core.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:2703: modules/dnn/CMakeFiles/opencv_test_dnn.dir/rule] Error 2
make: *** [Makefile:626: opencv_test_dnn] Error 2
```

### Steps to reproduce

```
cmake -DCPU_BASELINE="NEON;NEON_FP16;NEON_DOTPROD" ../opencv
make -j4
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

### PR DESCRIPTION
**Pull request by tomoaki0705**
closes #24588

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=ARMv7,ARMv8
```

### CODE REVIEW

**Comment by tomoaki0705** [comment]
OK, now I checked several configuration and basically, it seems fine.


|Architecture|Compiler|`CPU_BASELINE`|compiler flag at the end|
|---|---|---|---|
|Aarch64|GCC 13.2|`"NEON_BF16;NEON_FP16;NEON_DOTPROD"`|`-march=armv8.2-a+dotprod+fp16+bf16`|
|Aarch64|GCC 13.2|`"NEON_BF16;NEON_DOTPROD"`|`-march=armv8.2-a+dotprod+bf16`|
|Aarch64|GCC 13.2|`"NEON_FP16;NEON_DOTPROD"`|`-march=armv8.2-a+dotprod+fp16`|
|Aarch64|GCC 13.2|`"NEON_FP16;NEON_BF16"`|`-march=armv8.2-a+fp16+bf16`|
|Aarch64|GCC 13.2|`NEON_BF16`|`-march=armv8.2-a+bf16`|
|Aarch64|GCC 13.2|`NEON_FP16`|`-march=armv8.2-a+fp16`|
|Aarch64|GCC 13.2|`NEON_DOTPROD`|`-march=armv8.2-a+dotprod`|
|Aarch64|GCC 13.2|(none)|(none)|
|Aarch64|GCC 9.4|(none)|(none)|
|Aarch64|GCC 9.4|`"NEON_BF16;NEON_FP16;NEON_DOTPROD"`|`-march=armv8.2-a+dotprod+fp16 `|
|Aarch64|GCC 7.5|(none)|(none)|
|Aarch64|GCC 7.5|`"NEON_BF16;NEON_FP16;NEON_DOTPROD"`|`-march=armv8.2-a+dotprod+fp16 `|
|Aarch64|GCC 5.4|(none)|(none)|
|Aarch64|GCC 5.4|`"NEON_BF16;NEON_FP16;NEON_DOTPROD"`|(none)|
|Armv7|GCC 6.3|(none)|(none)|

- `bf16` support was introduced from 10 series and later.  It's expected that  `+bf16` doesn't appear at the end on 9 series and earlier compilers

```
-- Performing Test HAVE_CXX_MARCH_ARMV8_2_A+BF16 (check file: cmake/checks/cpu_neon_bf16.cpp)
-- Performing Test HAVE_CXX_MARCH_ARMV8_2_A+BF16 - Failed
-- NEON_BF16 is not supported by C++ compiler
-- Optimization NEON_BF16 is not available, skipped
-- Dispatch optimization NEON_BF16 is not available, skipped
```
- For GCC 5.4, it's expected that since `-march=armv8.2-a` option is supported from 7 series and later
- The last row (Armv7) is to show it doesn't fail in other platforms, but it's guarded with `if(AARCH64)` so it should not harm the other platform

File: cmake/OpenCVCompilerOptimizations.cmake [hunk 1]
```diff
@@ -480,6 +480,23 @@ macro(ocv_check_compiler_optimization OPT)
   endif()
 endmacro()
 
+macro(merge_feature_options FEATURE_NAME_LIST FLAG_STRING COMMON_OPTION)
```
Comment (opencv-alalek) [inline cmake/OpenCVCompilerOptimizations.cmake:483]: `ocv_cpu_aarch64_baseline_` prefix

Comment (tomoaki0705) [reply cmake/OpenCVCompilerOptimizations.cmake]: Thank you!
Done.

**Review by asmorkalov — APPROVED** [review]
:+1:

**Review by vpisarev — APPROVED** [review]

**Pull request merged**